pkg/identity/service.go

```diff
@@ -0,0 +1,187 @@
+package identity
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/linkerd/linkerd2/pkg/tls"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	pb "github.com/linkerd/linkerd2-proxy-api/go/identity"
+)
+
+const (
+	// DefaultIssuanceLifetime is the default lifetime of certificates issued by
+	// the identity service.
+	DefaultIssuanceLifetime = 24 * time.Hour
+)
+
+type (
+	// Service implements the gRPC service in terms of a Validator and Issuer.
+	Service struct {
+		v Validator
+		i tls.Issuer
+	}
+
+	// Validator implementors accept a bearer token, validates it, and returns a
+	// DNS-form identity.
+	Validator interface {
+		// Validate takes an opaque authentication token, attempts to validate its
+		// authenticity, and produces a DNS-like identifier.
+		//
+		// An InvalidToken error should be returned if the provided token was not in a
+		// correct form.
+		//
+		// A NotAuthenticated error should be returned if the authenticity of the
+		// token cannot be validated.
+		Validate(context.Context, []byte) (string, error)
+	}
+
+	// InvalidToken is an error type returned by Validators to indicate that the
+	// provided authentication token was not valid.
+	InvalidToken struct{ Reason string }
+
+	// NotAuthenticated is an error type returned by Validators to indicate that the
+	// provided authentication token could not be authenticated.
+	NotAuthenticated struct{}
+)
+
+// NewService creates a new identity service.
+func NewService(v Validator, i tls.Issuer) *Service {
+	return &Service{v, i}
+}
+
+// Register registers an identity service implementation in the provided gRPC
+// server.
+func Register(g *grpc.Server, s *Service) {
+	pb.RegisterIdentityServer(g, s)
```

Comment:
This doesn't compile for me. Our pattern in most of our other gRPC servers is to include a `NewServer` initialization func that handles registration and returns `*grpc.Server` and `net.Listener`. For instance:

https://github.com/linkerd/linkerd2/blob/81f645da665bcec99ccfc26233d7d84a56ce41f3/controller/tap/server.go#L430-L452

Can we do something like that here?

Reply:
This pkg module is intentionally kubernetes-agnostic; and it exposes this helper so that the caller doesn't need to import the protobuf library just for this one call.

It would be a shame to make this module pull in a kubernetes dependency just for this builder. But perhaps the controller module should export a constructor like this?

Reply:
The poiny about kubernetes isn't exactly relevant I guess.. but the larger point is that the `pkg/` module is just a "library" and doesn't have any grpc.Server/net.Listener dependencies -- it only deals in the interfaces.  I think the `controller/identity` package should be responsible for this, if we care to split out of the main...

Reply:
My thought process is the following: pkg/identity implements the grpc interface in terms of two other interfaces: Validator, and Issuer.  The controller, which handles configuration and runtime logic, builds a validator and Issuer.  I structured this specifically this way because I know of another user that wants to implement another identity service. Providing a clear library interface for implementing this seemed preferable to bundling the logical service and the concrete server.

Reply:
Ok, that all makes sense. Since this implementation deviates from the implementation of the rest of our gRPC services in the project, it's adding complexity to the codebase. We can certainly justify the complexity, as it exists to satisfy some external use case, but I wanted to at least note it.